WooCommerce/Classes/ViewRelated/Orders/Order Details/Issue Refunds/OrderRedundsOptionsDeterminer.swift

```diff
@@ -0,0 +1,78 @@
+import Foundation
+import Yosemite
+
+protocol OrderRefundsOptionsDeterminerProtocol {
+    func determineRefundableOrderItems(from order: Order, with refunds: [Refund]) -> [RefundableOrderItem]
+    func isAnythingToRefund(from order: Order, with refunds: [Refund], currencyFormatter: CurrencyFormatter) -> Bool
+}
+
+final class OrderRefundsOptionsDeterminer: OrderRefundsOptionsDeterminerProtocol {
+
+    func isAnythingToRefund(from order: Order, with refunds: [Refund], currencyFormatter: CurrencyFormatter) -> Bool {
+        var shippingLineCanBeRefunded = false
+        if let shippingLine = order.shippingLines.first {
+            shippingLineCanBeRefunded = shippingLineIsRefundable(shippingLine, currencyFormatter: currencyFormatter)
+
+        }
+
+        return isAnyFeeAvailableForRefund(from: order) ||
+        hasShippingBeenRefunded(from: refunds) == false ||
+        shippingLineCanBeRefunded ||
+        determineRefundableOrderItems(from: order, with: refunds).count > 0
+    }
+
+    /// Return an array of `RefundableOrderItems` by taking out all previously refunded items
+    ///
+    func determineRefundableOrderItems(from order: Order, with refunds: [Refund]) -> [RefundableOrderItem] {
+        // Flattened array with all items refunded
+        let allRefundedItems = refunds.flatMap { $0.items }
+
+        // Transform `order.items` by subtracting the quantity left to refund and evicting those who were fully refunded.
+        return order.items.compactMap { item -> RefundableOrderItem? in
+
+            // Calculate how many times an item has been refunded. This number is negative.
+            let timesRefunded = allRefundedItems.reduce(0) { timesRefunded, refundedItem -> Decimal in
+
+                // Only keep accumulating if the refunded item product and the original item product match
+                guard refundedItem.productOrVariationID == item.productOrVariationID else {
+                    return timesRefunded
+                }
+                return timesRefunded + refundedItem.quantity
+            }
+
+            // If there is no more items to refund, evict it from the resulting array
+            let quantityLeftToRefund = item.quantity + timesRefunded
+            guard quantityLeftToRefund > 0 else {
+                return nil
+            }
+
+            // Return the item with the updated quantity to refund
+            return RefundableOrderItem(item: item, quantity: quantityLeftToRefund)
+        }
+    }
+
+    func isAnyFeeAvailableForRefund(from order: Order) -> Bool {
```

Comment:
nit: maybe this and other helpers below (`hasShippingBeenRefunded`, `shippingLineIsRefundable`) can be moved to a private extension, or marked as private?

Reply:
As I refactored the logic, this is no relevant anymore 😃